Fixes part of #159: UI Structure - QuestionTrainingFragment (DO NOT MERGE) [Blocked: #397, #495]

jamesxu0 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes part of #159: UI Structure - QuestionTrainingFragment (DO NOT MERGE). This PR was made for @BenHenning to branch off of to finish the QuestionAssessmentProgressController.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch.
- [ ] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

jamesxu0 committed:
Merge branch 'introduce-question-progress-controller-interface' into question_player_lofi

jamesxu0 committed:
Started UI

jamesxu0 committed:
setup ViewModel

jamesxu0 committed:
added binding adapter

jamesxu0 committed:
Handles basic functionality

jamesxu0 committed:
Moved code around

jamesxu0 committed:
Added replay button

jamesxu0 committed:
Merge branch 'develop' into question_player_lofi

BenHenning committed:
Merge branch 'develop' into question_player_lofi

Conflicts:
	app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
	app/src/main/res/layout/question_player_fragment.xml
	app/src/sharedTest/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivityTest.kt
	domain/src/main/java/org/oppia/domain/question/QuestionTrainingController.kt
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into question_player_lofi

Conflicts:
	domain/src/test/java/org/oppia/domain/question/QuestionAssessmentProgressControllerTest.kt

BenHenning committed:
Make questions partially work.

This introduces back button behavior to ensure training sessions can be
reset.

This also attempts to introduce proper asset loading, but it doesn't
seem to be working currently.

BenHenning committed:
Temporarily remove check in StateDeck that breaks questions.

BenHenning commented:
NB: This is branch is now a functional version of questions. It seems like some code needs to be cleaned up & restructured to encourage more sharing. #397 also needs to be finished (there's one bug to fix, and its tests need to be finished).

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into question_player_lofi

Conflicts:
	app/src/main/res/values/colors.xml

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

BenHenning committed:
Post-merge fixes with some temporary stop-gap solutions.

BenHenning committed:
Fix questions image rendering and consolidate GCS resource bucket names.

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

BenHenning committed:
Fix index accounting for questions. Before, the index incremented upon
correct answer submission rather than navigation.

BenHenning committed:
Remove support for backward navigation in questions since it isn't
allowed.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

Conflicts:
	domain/src/main/java/org/oppia/domain/state/StateDeck.kt

BenHenning committed:
Set title of question activity to 'Train Mode'.

BenHenning committed:
Add support for a synthetic ephemeral question to represent the end of a
training session.

BenHenning committed:
Merge branch 'introduce-question-progress-controller-implementation' into temp-base-branch-for-question-player

BenHenning committed:
Merge branch 'temp-base-branch-for-question-player' into question_player_lofi

Conflicts:
	domain/src/main/java/org/oppia/domain/question/QuestionAssessmentProgressController.kt
	model/src/main/proto/question.proto

BenHenning committed:
Update question progress tracking to be translatable and to ensure that
the synthetic terminal state shows the correct progress: 'completed'.

BenHenning committed:
Move most state player functional support into a new configurable
assembler class to enable sharing between the exploration and question
players.

BenHenning committed:
Hook up StatePlayerRecyclerViewAssembler to the question player fragment
presenter. This also enables the congratulations text.

BenHenning committed:
Introduce proper terminal page and disable overscroll. Note that the end
session page introduced here has not been mocked yet, so it's mostly a
placeholder until a mock is available.

BenHenning committed:
Add support & button for training session replay.

BenHenning commented:
This has been replaced by #500. Thanks for creating this @jamesxu0!
